Chapter 5 production polish (#78)
UBC-DSCI/introduction-to-datascience-python: source/classification1.md

  name: python3
---

```{code-cell} ipython3
:tags: [remove-cell]

> import warnings
> def warn(*args, **kwargs):
>     pass
> warnings.warn = warn

Reviewer: ```suggestion
import warnings
warnings.filterwarnings("ignore", category=DeprecationWarning)
```

Does this work for you? I left a note in https://github.com/UBC-DSCI/introduction-to-datascience-python/issues/80 too

Author: I tried this originally, and it didn't work. Let me try it again now.

Author: ah. Duh. the trick is to ignore `FutureWarning`, not `DeprecationWarning`. Will edit this and other chapters.

```{index} info
```

> Below we use the `info` method to preview the data frame. This method can 

Reviewer: I *think* we have been calling it "function" everywhere so far (for simplicity even if it is incorrect). Maybe something we can do a pass on at the end and see that we are consistent. I guess we have https://github.com/UBC-DSCI/introduction-to-datascience-python/issues/65 for this already

Author: Yep, I'll punt this to #65

as well as their data types and the number of non-missing entries.

```{code-cell} ipython3
cancer.info()
```

```{index} unique
```

From the summary of the data above, we can see that `Class` is of type `object`.
We can use the `unique` method on the `Class` column to see all unique values
present in that column. We see that there are two diagnoses:
benign, represented by `'B'`, and malignant, represented by `'M'`.

```{code-cell} ipython3
cancer['Class'].unique()
```

We will also improve the readability of our analysis
by renaming `'M'` to `'Malignant'` and `'B'` to `'Benign'` using the `replace`
method. The `replace` method takes one argument: a dictionary that maps
previous values to desired new values. 
Furthermore, since we will be working with `Class` as a categorical statistical variable,
it is a good idea to convert it to the `category` type using the `astype` method 
on the `cancer` data frame. We will verify the result using the `info` 
and `unique` methods again.

```{index} replace
```

```{code-cell} ipython3
cancer['Class'] = cancer['Class'].replace({
				'M' : 'Malignant',
				'B' : 'Benign'
				})

> cancer['Class'] = cancer['Class'].astype('category')

Reviewer: ```suggestion
cancer['Class'] = cancer['Class'].replace({
    'M' : 'Malignant',
    'B' : 'Benign'
}).astype('category')
```

I like the choice of `replace` instead of `map`. In addition to the NA issue you mentioned I think the method name if more intuitive. I think we should cover the `category` data type in the wrangling chapter so that students know why it is beneficial (although I also wonder if we will actually leverage any of the benefits in dsci100 or if it is just one more thing to learn, sorry I know I suggested it initially). https://github.com/UBC-DSCI/introduction-to-datascience-python/issues/120

Author: Let's stick with `category` -- that data type itself isn't such a huge deal, but the important lesson for students to learn is that they should care about data types generally.

I like your suggestion on the code block, will make this change everywhere it's relevant

cancer.info()
```

```{code-cell} ipython3
cancer['Class'].unique()
```

### Exploring the cancer data

```{index} groupby, count
```

```{code-cell} ipython3
:tags: [remove-cell]
glue("benign_count", cancer['Class'].value_counts()['Benign'])
glue("benign_pct", int(np.round(100*cancer['Class'].value_counts(normalize=True)['Benign'])))
glue("malignant_count", cancer['Class'].value_counts()['Malignant'])
glue("malignant_pct", int(np.round(100*cancer['Class'].value_counts(normalize=True)['Malignant'])))
```

Before we start doing any modeling, let's explore our data set. Below we use

> the `groupby` and `count` methods to find the number and percentage 
> of benign and malignant tumor observations in our data set. When paired with
> `groupby`, `count` counts the number of observations for each value of the `Class`
> variable. Then we calculate the percentage in each group by dividing by the total
> number of observations and multiplying by 100. We have 
> {glue:}`benign_count` ({glue:}`benign_pct`\%) benign and
> {glue:}`malignant_count` ({glue:}`malignant_pct`\%) malignant 
> tumor observations.

Reviewer: ```suggestion
the `groupby` and `size` methods to find the number and percentage 
of benign and malignant tumor observations in our data set. When paired with
`groupby`, `size` counts the number of observations for each value of the `Class`
variable.
```

Reviewer: I moved the rest down so that we do one step at a time

```{code-cell} ipython3

> explore_cancer = pd.DataFrame()
> explore_cancer['count'] = cancer.groupby('Class')['ID'].count()

Reviewer: ```suggestion
cancer.groupby('Class').size()
```

There is a `size` method which I think is more straightforward since we don't have to select a column. I don't see the value in creating a new data frame here and think it is easier to understand if we just do it in two different steps, as with value_counts below

> explore_cancer['percentage'] = 100 * explore_cancer['count']/len(cancer)
> explore_cancer
> ```

Reviewer: ````suggestion
```

Then we calculate the percentage in each group by dividing by the total
number of observations and multiplying by 100. We have 
{glue:}`benign_count` ({glue:}`benign_pct`\%) benign and
{glue:}`malignant_count` ({glue:}`malignant_pct`\%) malignant 
tumor observations.

```{code-cell} ipython3
100 * cancer.groupby('Class').size() / cancer.shape[0]
```
````

`shape` is preferred over length on dataframes since it is already computed and in the spirit of using dataframes methods instead of global python functions in general.

Author: Opened an issue about converting all to shape throughout

I will open another issue about size -- we may need to edit earlier chapters to do that

```{index} value_counts
```

The `pandas` package also has a more convenient specialized `value_counts` method for 
counting the number of occurrences of each value in a column. If we pass no arguments
to the method, it outputs a series containing the number of occurences
of each value. If we instead pass the argument `normalize=True`, it instead prints the fraction
of occurrences of each value.

```{code-cell} ipython3
cancer['Class'].value_counts()
```

```{code-cell} ipython3
cancer['Class'].value_counts(normalize=True)
```

```{index} visualization; scatter
```

Next, let's draw a colored scatter plot to visualize the relationship between the
perimeter and concavity variables. Recall that `altair's` default palette
is colorblind-friendly, so we can stick with that here.

```{code-cell} ipython3
perim_concav = (
    alt.Chart(cancer)

>     .mark_point(opacity=0.6, filled=True, size=40)

Reviewer: ```suggestion
    .mark_circle()
```

I think we should keep the anything that is not central to the chapter as simple as possible. In this case there is not much difference by adding these three parameters

New:
![image](https://user-images.githubusercontent.com/4560057/211333814-50bf1e1e-d8d5-4f5f-b936-407cfa0a18f6.png)

Old:
![image](https://user-images.githubusercontent.com/4560057/211333882-75d968c2-f8c1-481b-9924-7dcec151fbda.png)

To illustrate the concept of $K$-nearest neighbors classification, we 
will walk through an example.  Suppose we have a
new observation, with standardized perimeter 
of {glue:}`new_point_1_0` and standardized concavity 
of {glue:}`new_point_1_1`, whose 
diagnosis "Class" is unknown. This new observation is 
depicted by the red, diamond point in {numref}`fig:05-knn-2`.

```{code-cell} ipython3
:tags: [remove-cell]

perim_concav_with_new_point = (
    alt.Chart(
        perim_concav_with_new_point_df,
    )

>     .mark_point(opacity=0.6, filled=True, size=40)

Reviewer: ```suggestion
    .mark_circle()
```

    .encode(
        x=alt.X("Perimeter", title="Perimeter (standardized)"),
        y=alt.Y("Concavity", title="Concavity (standardized)"),
        color=alt.Color("Class", title="Diagnosis"),
        shape=alt.Shape(
            "Class", scale=alt.Scale(range=["circle", "circle", "diamond"])
        ),

>         size=alt.condition("datum.Class == 'Unknown'", alt.value(80), alt.value(30)),

Reviewer: ```suggestion
        size=alt.condition("datum.Class == 'Unknown'", alt.value(100), alt.value(30)),
        stroke=alt.condition("datum.Class == 'Unknown'", alt.value('black'), alt.value(None)),
```
I vote for making this new point easier to see:

After this change:
![image](https://user-images.githubusercontent.com/4560057/211337575-ce739a60-2f15-475b-b5c9-3d20affbb930.png)

Previously:
![image](https://user-images.githubusercontent.com/4560057/211337365-caedfcca-3af8-40b0-8baa-8b51213d1508.png)

> glue('fig:05-knn-4', (perim_concav_with_new_point2 + line2), display=True)

Reviewer: ```suggestion
perim_concav_with_new_point['data'] = perim_concav_with_new_point_df2
line['data'] = new_neighbor_df2
glue('fig:05-knn-4', (perim_concav_with_new_point + line), display=True)
```

We don't need to recreate the charts each time with the exact same spec. It is enough to just switch out the data, so we can delete all the code for `line2` and `perim_concav_with_new_point2`; the same goes below. This will make it easier to keep this good up to date since changes only happen in one location.

I think it is harmless to overwrite the data in the same variable in this case, but if we don't want that for some reason, we can use `perim_concav_with_new_point.properties(data=perim_concav_with_new_point_df2)` to create a new chart, it's just a bit more verbose.

Author: Something was off about this... it didn't work for me. We can do this later, it's good enough for now as is

```{code-cell} ipython3
new_obs_Perimeter = 0
new_obs_Concavity = 3.5
cancer_dist = (cancer
           .loc[:, ["Perimeter", "Concavity", "Class"]]
           .assign(dist_from_new = (
               (cancer["Perimeter"] - new_obs_Perimeter) ** 2
             + (cancer["Concavity"] - new_obs_Concavity) ** 2
           )**(1/2))
           .nsmallest(5, "dist_from_new")
      )

> cancer_dist

Reviewer: ```suggestion
(
    cancer
   [["Perimeter", "Concavity", "Class"]]
   .assign(dist_from_new = (
       (cancer["Perimeter"] - new_obs_Perimeter) ** 2
     + (cancer["Concavity"] - new_obs_Concavity) ** 2
   )**(1/2))
   .nsmallest(5, "dist_from_new")
)
```

Remove redundant assignment and unnecessary use of loc

new_obs_Perimeter = 0
new_obs_Concavity = 3.5
new_obs_Symmetry = 1
cancer_dist2 = (cancer
           .loc[:, ["Perimeter", "Concavity", "Symmetry", "Class"]]
           .assign(dist_from_new = (
               (cancer["Perimeter"] - new_obs_Perimeter) ** 2
             + (cancer["Concavity"] - new_obs_Concavity) ** 2
             + (cancer["Symmetry"] - new_obs_Symmetry) ** 2
           )**(1/2))
           .nsmallest(5, "dist_from_new")
      )

> cancer_dist2

Reviewer: ```suggestion
(
    cancer
    [["Perimeter", "Concavity", "Symmetry", "Class"]]
    .assign(dist_from_new = (
        (cancer["Perimeter"] - new_obs_Perimeter) ** 2
        + (cancer["Concavity"] - new_obs_Concavity) ** 2
        + (cancer["Symmetry"] - new_obs_Symmetry) ** 2
    )**(1/2))
    .nsmallest(5, "dist_from_new")
)
```

and a standard deviation of 1. To illustrate the effect that standardization can have on the $K$-nearest
neighbor algorithm, we will read in the original, unstandardized Wisconsin breast
cancer data set; we have been using a standardized version of the data set up
until now. We will apply the same initial wrangling steps as we did earlier,
and to keep things simple we will just use the `Area`, `Smoothness`, and `Class`
variables:

```{code-cell} ipython3
unscaled_cancer = (
            pd.read_csv("data/unscaled_wdbc.csv")
            .loc[:, ['Class', 'Area', 'Smoothness']]
            .replace({
               'M' : 'Malignant',
               'B' : 'Benign'
            })
    )

> unscaled_cancer['Class'] = unscaled_cancer['Class'].astype('category')

Reviewer: ```suggestion
unscaled_cancer = pd.read_csv("data/unscaled_wdbc.csv")[['Class', 'Area', 'Smoothness']]
unscaled_cancer['Class'] = unscaled_cancer['Class'].replace({
   'M' : 'Malignant',
   'B' : 'Benign'
}).astype('category')
unscaled_cancer
```

I think we should be careful showing replace on the entire dataframe without talking about it explicitly. It is easy to make a mistake and replace values in columns that were not intended to be replaced, so it's safer to limit it to one column.

```{index} scikit-learn; fit, scikit-learn; transform
```

You can see that the preprocessor includes a single standardization step
that is applied to the `Area` and `Smoothness` columns. 
Note that here we specified which columns to apply the preprocessing step to 
by individual names; this approach can become quite difficult, e.g., when we have many
predictor variables. Rather than writing out the column names individually,
we can instead use the 
[`make_column_selector`](https://scikit-learn.org/stable/modules/generated/sklearn.compose.make_column_selector.html#sklearn.compose.make_column_selector) function. For example, if we wanted to standardize all *numerical* predictors,
we would use `make_column_selector` and specify the `dtype_include` argument to be `np.number`
(from the `numpy` package). This creates a preprocessor equivalent to the one we
created previously.

```{code-cell} ipython3
import numpy as np
from sklearn.compose import make_column_selector

preprocessor = make_column_transformer(

>            (StandardScaler(), make_column_selector(dtype_include=np.number)),

Reviewer: Note that we could simply use `'number'` here instead of importing numpy. The drawback is that there is no error directly in this step if there is a typo so it could be confusing. The upside is that we could specify `'category'` the same way to select all pandas categories. See the notes section here https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.select_dtypes.html#pandas.DataFrame.select_dtypes

Author: I like your suggestion over the numpy version. Will change.

>            (StandardScaler(), make_column_selector(dtype_include=np.number)),

Reviewer: ```suggestion
    (StandardScaler(), make_column_selector(dtype_include=np.number)),
```

>       )

Reviewer: ```suggestion
)
```

Reviewer: In case this large indentation on open params happens because of the default vim settings, which was something that was driving me nuts for a long while, you can set `let g:pyindent_open_paren = 'shiftwidth()'` as per https://stackoverflow.com/questions/23016531/how-to-make-vim-only-insert-4-spaces-one-indent-level-for-the-next-line-after/74711118#74711118

This enables us to compute the quantities needed to standardize using one data set, and then 
apply that standardization to another data set.

```{code-cell} ipython3
preprocessor.fit(unscaled_cancer)
scaled_cancer = preprocessor.transform(unscaled_cancer)
scaled_cancer
```
```{code-cell} ipython3

```
It looks like our `Smoothness` and `Area` variables have been standardized. Woohoo!
But there are two important things to notice about the new `scaled_cancer` data frame. First, it only keeps
the columns from the input to `transform` (here, `unscaled_cancer`) that had a preprocessing step applied
to them. The default behavior of the `ColumnTransformer` that we build using `make_column_transformer` 
is to *drop* the remaining columns. This default behavior works well with the rest of `sklearn` (as we will see below
in the {ref}`08:puttingittogetherworkflow` section), but for visualizing the result of preprocessing it can be useful to keep the other columns
in our original data frame, such as the `Class` variable here.
To keep other columns, we need to set the `remainder` argument to `'passthrough'` in the `make_column_transformer` function.

>  Furthermore, you can see that the new column names---{glue:}`scaled_cancer.columns[0]`

Reviewer: The glued variables are now showing up for me

![image](https://user-images.githubusercontent.com/4560057/211351596-b4c2c68e-d022-4350-85ef-ff156435a105.png)

Author: fixed

and {glue:}`scaled_cancer.columns[1]`---include the name
of the preprocessing step separated by underscores. This default behavior is useful in `sklearn` because we sometimes want to apply
multiple different preprocessing steps to the same columns; but again, for visualization it can be useful to preserve
the original column names. To keep original column names, we need to set the `verbose_feature_names_out` argument to `False`.

> **Note:** Only specify the `remainder` and `verbose_feature_names_out` arguments when you want to examine the result
> of your preprocessing step. In most cases, you should leave these arguments at their default values.

```{code-cell} ipython3

> preprocessor_keep_all = make_column_transformer(
>            (StandardScaler(), make_column_selector(dtype_include=np.number)),
>             remainder='passthrough',
>             verbose_feature_names_out=False
>       )

Reviewer: ```suggestion
preprocessor_keep_all = make_column_transformer(
    (StandardScaler(), make_column_selector(dtype_include=np.number)),
    remainder='passthrough',
    verbose_feature_names_out=False
)
```

picking only 3 observations from the malignant group, and keeping all
of the benign observations. We choose these 3 observations using the `.head()`
method, which takes the number of rows to select from the top (`n`).
We use the [`concat`](https://pandas.pydata.org/docs/reference/api/pandas.concat.html) 
function from `pandas` to glue the two resulting filtered
data frames back together by passing them together in a sequence.
The new imbalanced data is shown in {numref}`fig:05-unbalanced`, 
and we print the counts of the classes using the `value_counts` function.

```{code-cell} ipython3

> rare_cancer = pd.concat(

Reviewer: ```suggestion
rare_cancer = pd.concat((
```

>     (cancer[cancer["Class"] == 'Benign'],
>      cancer[cancer["Class"] == 'Malignant'].head(3)
>     ))

Reviewer: ```suggestion
    cancer[cancer["Class"] == 'Benign'],
    cancer[cancer["Class"] == 'Malignant'].head(3)
))
```

Nice this new approach is much more intuitive than the one with `apply`! I moved this parent up one line

rare_plot = (
    alt.Chart(rare_cancer)

>     .mark_point(opacity=0.6, filled=True, size=40)

Reviewer: ```suggestion
    .mark_circle()
```

                domain=(rare_cancer["Concavity"].min(), rare_cancer["Concavity"].max())
            ),
        ),
        color=alt.Color("Class", title="Diagnosis"),
    )
)
#rare_plot + prediction_plot

> glue("fig:05-upsample-2", (rare_plot + prediction_plot))

Reviewer: Hmm, for some reason I see the wrong chart in the book I build locally here:

![image](https://user-images.githubusercontent.com/4560057/211354761-3c6074d0-02c4-4ec0-9512-92fb50eac85c.png)

Ah, the reason is that the glue variable is overwritten in the last chart. see my edit just before the exercises for fixing this

Author: yep, I fixed that just now before looking at your review :)

fairly nuanced, and a careful treatment would require a lot more detail and mathematics than we will cover in this textbook.
For the present purposes, it will suffice to rebalance the data by *oversampling* the rare class. 
In other words, we will replicate rare observations multiple times in our data set to give them more
voting power in the $K$-nearest neighbor algorithm. In order to do this, we will 
first separate the classes out into their own data frames by filtering.
Then, we will
use the [`resample`](https://scikit-learn.org/stable/modules/generated/sklearn.utils.resample.html) function 
from the `sklearn` package to increase the number of `Malignant` observations to be the same as the number 
of `Benign` observations. We set the `n_samples` argument to be the number of `Malignant` observations we want. 
We also set the `random_state` to be some integer
so that our results are reproducible; if we do not set this argument, we will get a different upsampling each time
we run the code. Finally, we use the `value_counts` method

>  to see that our classes are now balanced.

Reviewer: Here we introduce random state, then in the next chapter we talk about seeds instead.

Author: good point. will edit that to remove mention and hide a call to `np.random.seed` internally.

+++

(08:puttingittogetherworkflow)=
## Putting it together in a `Pipeline`

```{index} scikit-learn; pipeline
```

The `scikit-learn` package collection also provides the [`Pipeline`](https://scikit-learn.org/stable/modules/generated/sklearn.pipeline.Pipeline.html?highlight=pipeline#sklearn.pipeline.Pipeline), 
a  way to chain together multiple data analysis steps without a lot of otherwise necessary code for intermediate steps.
To illustrate the whole workflow, let's start from scratch with the `unscaled_wdbc.csv` data.
First we will load the data, create a model, and specify a preprocessor for the data.

```{code-cell} ipython3
# load the unscaled cancer data, make Class readable

> unscaled_cancer = (
>             pd.read_csv("data/unscaled_wdbc.csv")
>             .replace({
>                'M' : 'Malignant',
>                'B' : 'Benign'
>             })
>     )
> # make Class a categorical type
> unscaled_cancer['Class'] = unscaled_cancer['Class'].astype('category')

Reviewer: ```suggestion
unscaled_cancer = pd.read_csv("data/unscaled_wdbc.csv")
unscaled_cancer['Class'] = unscaled_cancer['Class'].replace({
   'M' : 'Malignant',
   'B' : 'Benign'
}).astype('category')
unscaled_cancer
```

step drops all the variables except the two we listed: `Area` and `Smoothness`.
For the `y` response variable argument, we pass the `unscaled_cancer["Class"]` series as before.

```{code-cell} ipython3
from sklearn.pipeline import make_pipeline

knn_fit = make_pipeline(preprocessor, knn).fit(
    X=unscaled_cancer, 
    y=unscaled_cancer["Class"]
)

knn_fit
```

As before, the fit object lists the function that trains the model. But now the fit object also includes information about
the overall workflow, including the standardization preprocessing step.
In other words, when we use the `predict` function with the `knn_fit` object to make a prediction for a new
observation, it will first apply the same preprocessing steps to the new observation. 
As an example, we will predict the class label of two new observations:
one with `Area = 500` and `Smoothness = 0.075`, and one with `Area = 1500` and `Smoothness = 0.1`.

```{code-cell} ipython3
new_observation = pd.DataFrame({"Area": [500, 1500], "Smoothness": [0.075, 0.1]})
prediction = knn_fit.predict(new_observation)
prediction
```

The classifier predicts that the first observation is benign, while the second is

> malignant. {numref}`fig:05-workflow-plot-show` visualizes the predictions that this 

Reviewer 2: below uses `numpy.meshgrid`. I would be inclined to replace with `np.meshgrid` (or say "the `meshgrid` function from `numpy`)

Author: agree!

    .encode(
        x=alt.X("Area"),
        y=alt.Y("Smoothness"),
        color=alt.Color("Class", title="Diagnosis"),
    )
)
unscaled_plot + prediction_plot
```

```{code-cell} ipython3
:tags: [remove-input]

> glue("fig:05-upsample-2", (unscaled_plot + prediction_plot))

Reviewer: ```suggestion
glue("fig:05-upsample-3", (unscaled_plot + prediction_plot))
```

This overwrites a figure with the same name above. Maybe the figure name below should be changed too and use this one? Or where is this figure meant to be used?

Author: nah, I just caught the same issue and fixed it -- you're right, it was buggy.

there were actually some other buggy figs (color/point size was off) so I've fixed those too
